doc/user-manual/language/type-based-termination-checking.rst

```diff
@@ -0,0 +1,317 @@
+
+In our case, we use ``map`` and trust it. The evidence for this trust lies in the type signature of ``map: {A' B' : Set} → (A' → B') → List A' → List B'``. According to this signature, map operates on arbitrary types, meaning it cannot construct a malicious ``RoseTree``. This property, known as parametricity, aligns with the logic of Agda, as described in :ref:`[2]<type-based-termination-checking-references>`.
+
+The idea of the type-based termination checker is that Agda can try to track
+the "sizes" of terms, and deduce whether the recursive calls eventually happen on
+the terms of smaller sizes. In the case with the rose tree above, we see that ``map``
+operates on a type ``A' := RoseTree A``, where the type of arguments of ``mapRose f``
+is ``A'``, and the type of elements in ``children`` is also ``A'``. We can conclude
+that ``mapRose f``, if ever called, is invoked on the same elements as those in
+``children``, hence it makes a recursive call on a smaller argument than the enclosing function was initially called with.
+
+.. _type-based-termination-checking-usage:
+
+Usage
+-----
+
+To activate the type-based termination checker in Agda, you can use the option ``--type-based-termination``:
+
+::
+
+      {-# OPTIONS --type-based-termination #-}
+
+      id : {A : Set} → A → A
+      id x = x
+
+      plus : Nat → Nat → Nat
+      plus zero    m = m
+      plus (suc n) m = suc (plus (id n) m) -- note 'id' here
+
+Enabling this option is the only requirement to utilize the type-based termination checker.
+
+It is important to note that the type-based termination checker requires preprocessing of all datatypes involved in the checking process. For the provided definitions, the files containing ``Nat``, ``List``, and ``RoseTree`` should all be type-checked with the ``--type-based-termination`` option enabled.
+
+By default, type-based termination is disabled. You can also explicitly disable it by specifying ``--no-type-based-termination``.
+
+Additionally, there's an option ``--(no-)syntax-based-termination`` that enables or disables the default syntax-based termination checker. This option can be useful if you wish to exclusively experiment with the type-based termination.
+
+Another option is ``--(no-)size-preservation``, explained
+:ref:`here<type-based-termination-checking-size-preservation>`.
+
+For a deeper understanding of the internal behavior of the type-based termination checker, you can use the verbosity option ``term.tbt`` with different levels. An example of usage is ``{-# OPTIONS -v term.tbt:40 #-}`` in the file header. This option dumps the output of the checker. Note that Agda needs to be built in debug mode to print any verbose data.
```

Comment:
The `-v` options are normally reserved for developer use, so I would prefer to avoid mentioning them in the user manual. If there is a need to provide additional user-facing output, it would be better to introduce a new flag instead.

Reply:
I have a plan to write a tool that inserts Agda's sized types in the syntax according to the output of my algorithms. Currently, the debug output is highly cryptic, so until this tool is written, I'd rather remove this paragraph.